keanu-project/src/main/java/io/improbable/keanu/algorithms/sampling/ForwardSampler.java

```diff
@@ -0,0 +1,54 @@
+package io.improbable.keanu.algorithms.sampling;
+
+import io.improbable.keanu.KeanuRandom;
+import io.improbable.keanu.algorithms.NetworkSample;
+import io.improbable.keanu.algorithms.Variable;
+import io.improbable.keanu.algorithms.VariableReference;
+import io.improbable.keanu.algorithms.mcmc.SamplingAlgorithm;
+import io.improbable.keanu.vertices.NonProbabilistic;
+import io.improbable.keanu.vertices.Probabilistic;
+import io.improbable.keanu.vertices.Vertex;
+
+import java.util.List;
+import java.util.Map;
+
+public class ForwardSampler implements SamplingAlgorithm {
+
+    private static final double LOG_PROB_OF_PRIOR = 0.;
+
+    private final List<? extends Variable> variablesToSampleFrom;
+    private final List<Vertex> topologicallySortedVertices;
+    private final KeanuRandom random;
+
+    public ForwardSampler(List<? extends Variable> variablesToSampleFrom, List<Vertex> topologicallySortedVertices, KeanuRandom random) {
+        this.variablesToSampleFrom = variablesToSampleFrom;
+        this.topologicallySortedVertices = topologicallySortedVertices;
+        this.random = random;
+    }
+
+    @Override
+    public void step() {
+        for (Vertex vertex : topologicallySortedVertices) {
+            if (vertex instanceof Probabilistic) {
+                if (!(vertex.isObserved())) {
+                    vertex.setValue(((Probabilistic) vertex).sample(random));
+                }
+            } else if (vertex instanceof NonProbabilistic) {
+                vertex.setValue(((NonProbabilistic) vertex).calculate());
```

Comment:
the `setValue` is the only reason this class needs to know about `Vertex` rather than just dealing with `Variable`s. Is this the best design choice? I can't think of a better way...

Reply:
We can't get rid of `setValue` for this use case unfortunately